### PR TITLE
Migrate to kubebuilder v4 layout

### PR DESCRIPTION
**Pull request by rubenvp8510**

### CODE REVIEW

File: cmd/root/root.go [hunk 1]
```diff
@@ -45,25 +44,16 @@ func init() {
...
-			return fmt.Errorf("unable to load the config file: %w", err)
-		}
-	}
-
-	err = ctrlConfig.Validate()
+	ctrlCfg, options, err := LoadConfig(scheme, configFile)
 	if err != nil {
-		return fmt.Errorf("controller config validation failed: %w", err)
+		return err
 	}
```
[the remark this answers is not in the document]
Comment (rubenvp8510) [reply cmd/root/root.go:51]: I think this is fine, as we don't want to mask errors.

**Comment by rubenvp8510** [comment]
@iblancasa I've already address some of your comments 

@iblancasa @andreasgerstmayr Could you review this ? Thanks!

**Review by iblancasa — APPROVED** [review]

File: cmd/root/options.go [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+			"/debug/pprof/symbol":  pprof.Symbol,
+			"/debug/pprof/trace":   pprof.Trace,
+		}
+
+		if o.Metrics.ExtraHandlers == nil {
+			o.Metrics.ExtraHandlers = map[string]http.Handler{}
+		}
+
+		for path, handler := range endpoints {
+			o.Metrics.ExtraHandlers[path] = handler
```
Comment (andreasgerstmayr) [inline cmd/root/options.go:82]: I'm not sure if I fully understand this part, does this always enable the pprof endpoint if operator metrics are enabled?

If so, why? imho pprof should be enabled explicitly, not always when metrics are enabled.

**Comment by andreasgerstmayr** [comment]
lgtm overall, great work!
I left one comment regarding the pprof endpoints.

File: cmd/root/options.go (hunk 2 above)
Comment (rubenvp8510) [reply cmd/root/options.go:82]: that is actually a good point. I'll review this part and see what is the appropiate behavriour, I think we can add a flag ,

Comment (rubenvp8510) [reply cmd/root/options.go]: Just removed the pprof endpoint for now.

**Comment by andreasgerstmayr** [comment]
Last thing before merging, did you run the OpenShift e2e tests and confirm that they're still working?
Because there are some changes related to the metrics (.Metrics.BindAddress) in this PR, we should make sure that the operator and operand metrics still work, and iirc they are only tested in the OpenShift e2e test suite.

**Comment by rubenvp8510** [comment]
Pass almost all OCP tests except some due must gather issues

-- FAIL: chainsaw (398.06s)
    --- PASS: chainsaw/monitoring (124.40s)
    --- PASS: chainsaw/monolithic-monitoring (76.25s)
    --- PASS: chainsaw/red-metrics (197.41s)
    --- FAIL: chainsaw/monolithic-route (60.15s)
    --- PASS: chainsaw/monolithic-single-tenant-auth (72.51s)
    --- PASS: chainsaw/tls-singletenant-monolithic (77.18s)
    --- PASS: chainsaw/monolithic-multitenancy-static (79.68s)
    --- PASS: chainsaw/monolithic-multitenancy-openshift (80.40s)
    --- PASS: chainsaw/tls-singletenant (125.31s)
    --- PASS: chainsaw/tempo-single-tenant-auth (132.51s)
    --- FAIL: chainsaw/route (139.00s)
    --- PASS: chainsaw/multitenancy (145.20s)
    --- PASS: chainsaw/tempostack-resources (178.05s)
    --- PASS: chainsaw/component-replicas (202.22s

I don't understand why this step fails

**Comment by andreasgerstmayr** [comment]
> ```
>         Missing: *sha*/olm/operator-tempo-*-tempo-operator.yaml
>     | 00:38:13 | monolithic-route | Run the must-gather and verify the contents | SCRIPT    | ERROR |
> ```
> 
> @IshwarKanse Could you help me testing this on your side? May be is something on my environment. Thanks

I think the namespace where the operator is installed must end in `tempo-operator`. Does it work with the instructions here: https://github.com/grafana/tempo-operator/blob/main/RELEASE.md#running-e2e-tests-on-openshift ?

**Comment by IshwarKanse** [comment]
@andreasgerstmayr Yeah, we use openshift-tempo-operator namespace in CI. https://github.com/openshift/release/blob/master/ci-operator/config/openshift/grafana-tempo-operator/openshift-grafana-tempo-operator-main__upstream-ocp-4.16-amd64-aws-ipi-shared-vpc-phz-sts.yaml#L62 https://github.com/openshift/release/blob/master/ci-operator/step-registry/distributed-tracing/tests/tempo/upstream/distributed-tracing-tests-tempo-upstream-commands.sh I'll check this next week once I'm back from leave.

File: Makefile [hunk 3]
```diff
@@ -488,7 +488,7 @@ docs/spec/%: bundle/community/manifests/% | gen-api-docs
 	$(GEN_API_DOCS) < $^ > $@
 
 docs/operator/config.yaml: gen-api-docs
-	$(GEN_API_DOCS) -pkg github.com/grafana/tempo-operator/apis/config/v1alpha1 -type ProjectConfig -format multiline > $@
+	$(GEN_API_DOCS) -pkg github.com/grafana/tempo-operator/api/config/v1alpha1 -type ProjectConfig -format multiline > $@
```
Comment (pavolloffay) [inline Makefile:491]: Can you explain this change? Maybe a link to documentation? 

I think the `apis` was used for multi version projects.

Comment (pavolloffay) [reply Makefile:491]: and our PROJECT uses multigroup

File: PROJECT [hunk 4]
```diff
@@ -1,6 +1,6 @@
 domain: grafana.com
 layout:
-- go.kubebuilder.io/v3
+- go.kubebuilder.io/v4
```
Comment (pavolloffay) [inline PROJECT:3]: Can you point me to the docs for this file?

Comment (rubenvp8510) [reply PROJECT:3]: Sure, you can see the detailed steps here:

https://book.kubebuilder.io/migration/manually_migration_guide_gov3_to_gov4

Step: [Migrate the layout version into the PROJECT file](https://book.kubebuilder.io/migration/manually_migration_guide_gov3_to_gov4#migrate-the-layout-version-into-the-project-file)

File: Makefile (hunk 3 above)
Comment (rubenvp8510) [reply Makefile:491]: Not anymore. is part of the layout changes

[New layout:](https://book.kubebuilder.io/migration/manually_migration_guide_gov3_to_gov4#new-layout)

    The directory apis was renamed to api to follow the standard
    The controller(s) directory has been moved under a new directory called internal and renamed to singular as well controller
    The main.go previously scaffolded in the root directory has been moved under a new directory called cmd

**Comment by IshwarKanse** [comment]
> > ```
> >         Missing: *sha*/olm/operator-tempo-*-tempo-operator.yaml
> >     | 00:38:13 | monolithic-route | Run the must-gather and verify the contents | SCRIPT    | ERROR |
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > @IshwarKanse Could you help me testing this on your side? May be is something on my environment. Thanks
> 
> I think the namespace where the operator is installed must end in `tempo-operator`. Does it work with the instructions here: https://github.com/grafana/tempo-operator/blob/main/RELEASE.md#running-e2e-tests-on-openshift ?

@andreasgerstmayr I installed the operator in openshift-tempo-operator project and ran the tests. The must-gather steps pass. 

```
--- PASS: chainsaw (2294.18s)
    --- PASS: chainsaw/ossm-monolithic-otel (386.10s)
    --- PASS: chainsaw/ossm-tempostack (349.65s)
    --- PASS: chainsaw/ossm-tempostack-otel (374.55s)
    --- PASS: chainsaw/otel-tempo-serverless (412.65s)
    --- PASS: chainsaw/tempo-serverless (314.67s)
    --- PASS: chainsaw/monitoring (159.00s)
    --- PASS: chainsaw/monolithic-monitoring (101.48s)
    --- PASS: chainsaw/red-metrics (196.09s)
    --- PASS: chainsaw/operator-metrics (20.92s)
    --- PASS: chainsaw/reconcile (106.44s)
    --- PASS: chainsaw/monolithic-receivers-tls (114.84s)
    --- PASS: chainsaw/monolithic-pv (95.61s)
    --- PASS: chainsaw/monolithic-s3-tls (122.56s)
    --- PASS: chainsaw/tempostack-extraconfig (132.30s)
    --- PASS: chainsaw/receivers-tls (152.79s)
    --- PASS: chainsaw/receivers-mtls (153.01s)
    --- PASS: chainsaw/monolithic-extraconfig (54.30s)
    --- PASS: chainsaw/monolithic-memory (81.23s)
    --- PASS: chainsaw/monolithic-ingestion-mtls (74.47s)
    --- PASS: chainsaw/monolithic-route (76.84s)
    --- PASS: chainsaw/monolithic-single-tenant-auth (82.56s)
    --- PASS: chainsaw/generate (71.76s)
    --- PASS: chainsaw/gateway (43.53s)
    --- PASS: chainsaw/route (156.66s)
    --- PASS: chainsaw/monolithic-multitenancy-static (92.42s)
    --- PASS: chainsaw/multitenancy (162.88s)
    --- PASS: chainsaw/tempo-single-tenant-auth (143.93s)
    --- PASS: chainsaw/monolithic-multitenancy-openshift (80.69s)
    --- PASS: chainsaw/custom-ca (125.01s)
    --- PASS: chainsaw/tls-singletenant-monolithic (95.24s)
    --- PASS: chainsaw/compatibility (152.09s)
    --- PASS: chainsaw/tls-singletenant (157.65s)
    --- PASS: chainsaw/component-replicas (234.11s)
    --- PASS: chainsaw/tempostack-resources (197.49s)
    --- PASS: chainsaw/tempostack-retention-global (2862.40s)
PASS
Tests Summary...
- Passed  tests 35
- Failed  tests 0
- Skipped tests 0
Done.
```

**Review by andreasgerstmayr — APPROVED** [review]

**Pull request merged**